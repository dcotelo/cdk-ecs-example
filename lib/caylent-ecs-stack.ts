import * as cdk from '@aws-cdk/core';
import * as ec2 from "@aws-cdk/aws-ec2";
import * as ecs from "@aws-cdk/aws-ecs";
import * as ecs_patterns from "@aws-cdk/aws-ecs-patterns";
import { DockerImageAsset } from "@aws-cdk/aws-ecr-assets";
import * as dynamodb from "@aws-cdk/aws-dynamodb";
import CustomTable from './dynamodb/custom_table';



export class CaylentEcsStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // The code that defines your stack goes here
    const vpc = new ec2.Vpc(this, "MyVpc", {
      maxAzs: 3 // Default is all AZs in region
    });

    const cluster = new ecs.Cluster(this, "MyCluster", {
      vpc: vpc
    });


    //create and register image
    const myimage = new DockerImageAsset(this, "golang-example-app", {
      directory: 'golang-example-app/',
    });





    // Create a load-balanced Fargate service and make it public
    new ecs_patterns.ApplicationLoadBalancedFargateService(this, "MyFargateService", {
      cluster: cluster, // Required
      cpu: 512, // Default is 256
      desiredCount: 6, // Default is 1
      taskImageOptions: {
        //use our own image
        image: ecs.ContainerImage.fromDockerImageAsset(myimage),
        containerPort: 3000,
      },
      memoryLimitMiB: 2048, // Default is 512
      publicLoadBalancer: true, // Default is false,
    });

    //add custom table construct
    const prefix: string = "DEV"; //table prefix sholud be dynamic between environments
    const backupTag: string = "myBackupPlan"; // add table to a tag based backup plan

    const custom_table: CustomTable = new CustomTable(this, 'MY_CUSTOM_TABLE', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST
    }, prefix, backupTag);


  }
}
