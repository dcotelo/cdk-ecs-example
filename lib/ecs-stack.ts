import { Construct } from 'constructs';
import {Stack,StackProps,RemovalPolicy} from 'aws-cdk-lib';
import { Cluster,ContainerImage } from 'aws-cdk-lib/aws-ecs';
import { ApplicationLoadBalancedEc2Service, ApplicationLoadBalancedFargateService, } from 'aws-cdk-lib/aws-ecs-patterns';
import { DockerImageAsset } from 'aws-cdk-lib/aws-ecr-assets';
import { Vpc } from 'aws-cdk-lib/aws-ec2';
import { BillingMode } from 'aws-cdk-lib/aws-dynamodb';
import CustomTable from './dynamodb/custom_table';
import * as ec2 from 'aws-cdk-lib/aws-ec2';

enum ServiceType {
Fargate,
EC2
}


interface EcsClusterProps extends StackProps {
  dockerImageProp: DockerImageAsset,
  desiredCount:number,
  EcsServiceTypeProps?:ServiceType
}

export class EcsStack extends Stack {
  constructor(scope: Construct, id: string, props: EcsClusterProps) {
    super(scope, id, props);

    // The code that defines your stack goes here
    const vpc = new Vpc(this, "MyVpc", {
      maxAzs: 3 // Default is all AZs in region
    });

    const cluster = new Cluster(this, "MyCluster", {
      vpc: vpc,
      //enable container insights for better observability 
      containerInsights:true
    });


    if (props.EcsServiceTypeProps === ServiceType.Fargate) {
    // Create a load-balanced Fargate service and make it public
    new ApplicationLoadBalancedFargateService(this, "MyFargateService", {
      cluster: cluster, // Required
      cpu: 512, // Default is 256
      desiredCount: props.desiredCount, // Default is 1
      taskImageOptions: {
        //use our own image
        image: ContainerImage.fromDockerImageAsset(props.dockerImageProp),
        containerPort: 3000,
      },
      memoryLimitMiB: 2048, // Default is 512
      publicLoadBalancer: true, // Default is false,
    });
    }

    if (props.EcsServiceTypeProps === ServiceType.EC2) {
      new ApplicationLoadBalancedEc2Service(this, "MyEc2Service", {
        cluster: cluster, // Required
        cpu: 512, // Default is 256
        desiredCount: props.desiredCount, // Default is 1
        taskImageOptions: {
          //use our own image
          image: ContainerImage.fromDockerImageAsset(props.dockerImageProp),
          containerPort: 3000,
        },
        memoryLimitMiB: 2048, // Default is 512
        publicLoadBalancer: true, // Default is false,
      });

      // Add capacity to it
      cluster.addCapacity('DefaultAutoScalingGroupCapacity', {
        instanceType: new ec2.InstanceType("t2.micro"),
        desiredCapacity: 3,
});
    }


    //add custom table construct
    const prefix: string = "DEV"; //table prefix sholud be dynamic between environments
    const backupTag: string = "myBackupPlan"; // add table to a tag based backup plan

    const custom_table: CustomTable = new CustomTable(this, 'MY_CUSTOM_TABLE', {
      removalPolicy: RemovalPolicy.DESTROY,
      billingMode: BillingMode.PAY_PER_REQUEST
    }, prefix, backupTag);


  }
}
