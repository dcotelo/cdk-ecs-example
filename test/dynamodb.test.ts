import { expect as expectCDK, matchTemplate, MatchStyle } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Ecs from '../lib/ecs-stack';
import '@aws-cdk/assert/jest';
import { DockerImageAsset } from "@aws-cdk/aws-ecr-assets";
import { SynthUtils } from '@aws-cdk/assert';

const app = new cdk.App();

const DynamoTestStack = new cdk.Stack(app, "MyDynamoTestStack");

// WHEN
const myCustomImage = new DockerImageAsset(DynamoTestStack, "golang-example-app", {
    directory: 'golang-example-app/',
  });
const stack = new Ecs.EcsStack(DynamoTestStack, 'MyDynamoTestStack',{dockerImageProp:myCustomImage});

test('DynamoDB - Snapshot', () => {
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('DynamoDB', () => {
    expect(stack).toHaveResource("AWS::DynamoDB::Table", {
    });
});