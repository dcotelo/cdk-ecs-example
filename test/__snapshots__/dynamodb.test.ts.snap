// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamoDB - Snapshot 1`] = `
Object {
  "Outputs": Object {
    "MYCUSTOMTABLETableARN3A1A8A02": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "DEVMYCUSTOMTABLE0351AA05",
          "Arn",
        ],
      },
    },
    "MYCUSTOMTABLETopicARN7FE402CF": Object {
      "Value": Object {
        "Ref": "MYCUSTOMTABLETopic652FA786",
      },
    },
  },
  "Resources": Object {
    "DEVMYCUSTOMTABLE0351AA05": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "MY_CUSTOM_TABLE_ID",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "MY_CUSTOM_TABLE_ID",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableName": "DEV_MY_CUSTOM_TABLE",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "MYCUSTOMTABLETopic652FA786": Object {
      "Properties": Object {
        "DisplayName": "DEV_MY_CUSTOM_TABLE_TOPIC",
        "Tags": Array [
          Object {
            "Key": "backup-plan",
            "Value": "myBackupPlan",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
    "MyCluster4C1BA579": Object {
      "Properties": Object {
        "ClusterSettings": Array [
          Object {
            "Name": "containerInsights",
            "Value": "enabled",
          },
        ],
      },
      "Type": "AWS::ECS::Cluster",
    },
    "MyVpcF9F0CA6F": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "MyVpcIGW5C4A4F63": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "MyVpcPrivateSubnet1DefaultRouteA8CDE2FA": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "MyVpcPublicSubnet1NATGatewayAD3400C1",
        },
        "RouteTableId": Object {
          "Ref": "MyVpcPrivateSubnet1RouteTable8819E6E2",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcPrivateSubnet1RouteTable8819E6E2": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcPrivateSubnet1RouteTableAssociation56D38C7E": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "MyVpcPrivateSubnet1RouteTable8819E6E2",
        },
        "SubnetId": Object {
          "Ref": "MyVpcPrivateSubnet1Subnet5057CF7E",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcPrivateSubnet1Subnet5057CF7E": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcPrivateSubnet2DefaultRoute9CE96294": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "MyVpcPublicSubnet2NATGateway91BFBEC9",
        },
        "RouteTableId": Object {
          "Ref": "MyVpcPrivateSubnet2RouteTableCEDCEECE",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcPrivateSubnet2RouteTableAssociation86A610DA": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "MyVpcPrivateSubnet2RouteTableCEDCEECE",
        },
        "SubnetId": Object {
          "Ref": "MyVpcPrivateSubnet2Subnet0040C983",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcPrivateSubnet2RouteTableCEDCEECE": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcPrivateSubnet2Subnet0040C983": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcPublicSubnet1DefaultRoute95FDF9EB": Object {
      "DependsOn": Array [
        "MyVpcVPCGW488ACE0D",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "RouteTableId": Object {
          "Ref": "MyVpcPublicSubnet1RouteTableC46AB2F4",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcPublicSubnet1EIP096967CB": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "MyVpcPublicSubnet1NATGatewayAD3400C1": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "MyVpcPublicSubnet1EIP096967CB",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "MyVpcPublicSubnet1SubnetF6608456",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "MyVpcPublicSubnet1RouteTableAssociation2ECEE1CB": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "MyVpcPublicSubnet1RouteTableC46AB2F4",
        },
        "SubnetId": Object {
          "Ref": "MyVpcPublicSubnet1SubnetF6608456",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcPublicSubnet1RouteTableC46AB2F4": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcPublicSubnet1SubnetF6608456": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcPublicSubnet2DefaultRoute052936F6": Object {
      "DependsOn": Array [
        "MyVpcVPCGW488ACE0D",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "RouteTableId": Object {
          "Ref": "MyVpcPublicSubnet2RouteTable1DF17386",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcPublicSubnet2EIP8CCBA239": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "MyVpcPublicSubnet2NATGateway91BFBEC9": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "MyVpcPublicSubnet2EIP8CCBA239",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "MyVpcPublicSubnet2Subnet492B6BFB",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "MyVpcPublicSubnet2RouteTable1DF17386": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcPublicSubnet2RouteTableAssociation227DE78D": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "MyVpcPublicSubnet2RouteTable1DF17386",
        },
        "SubnetId": Object {
          "Ref": "MyVpcPublicSubnet2Subnet492B6BFB",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcPublicSubnet2Subnet492B6BFB": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "MyDynamoTestStack/MyDynamoTestStack/MyVpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcVPCGW488ACE0D": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
